### PR TITLE
S3gof3r for upload download

### PR DESCRIPTION
**Pull request by didip**
This PR introduces `s3gof3r` library for downloading and uploading files. The details on why it is needed is explained in this issue: https://github.com/dogestry/dogestry/issues/56

In short, `s3gof3r` does everything we need:
* Checksum
* Retries
* Multipart stream upload/download, in parallel.

The library offers so much advantages, it fixes multiple issues all at once: #10, #17, #55, and potentially #43. I'd love to get this PR out as soon as possible.

As a side-effect, the performance of uploading large binary file is also improved because that's `s3gof3r` core feature.

One other thing this PR does is to fix issue #57. `dogestry push` now fails immediately when 1 file failed to upload. #57 is the other part of solution to #43.

cc @toffer @relistan

### CODE REVIEW

**Comment by intjonathan** [comment]
:heart_eyes:

**Comment by didip** [comment]
@toffer Oh, since `s3gof3r` does checksum internally, we don't even need to upload .sum file, I think.
That said, I am not doing anything about that on this PR, we can solve that separately.

**Comment by didip** [comment]
@toffer There seems to be configuration problem here, I am getting: "The AWS Access Key Id you provided does not exist in our records." error on the latest build intermittently.

File: Godeps/_workspace/src/code.google.com/p/gcfg/read.go [hunk 1]
```diff
@@ -9,8 +9,8 @@ import (
 )
 
 import (
-	"github.com/dogestry/dogestry/Godeps/_workspace/src/code.google.com/p/gcfg/scanner"
-	"github.com/dogestry/dogestry/Godeps/_workspace/src/code.google.com/p/gcfg/token"
+	"code.google.com/p/gcfg/scanner"
```
Comment (relistan) [inline Godeps/_workspace/src/code.google.com/p/gcfg/read.go]: Is this really still mainly hosted on code.google.com? It's EOL

**Comment by relistan** [comment]
Wow, surprisingly few changes to get this done. :+1: Other than query above.

File: Godeps/_workspace/src/code.google.com/p/gcfg/read.go (hunk 1 above)
Comment (didip) [inline Godeps/_workspace/src/code.google.com/p/gcfg/read.go]: Yup, that's unfortunate.

**Comment by didip** [comment]
Please don't merge this PR yet, there are a few odd behaviors still.

**Comment by didip** [comment]
OK, now it is ready. @toffer @relistan @dselans

**Comment by dselans** [comment]
Nicely done!

**Comment by didip** [comment]
Testing instruction for those who are interested:
```
cd $GOPATH/go/src/github.com/dogestry/dogestry
git checkout s3gof3r-for-upload-download
docker pull giant-image:latest

# Test push
godep go run main.go push s3://$YOUR_S3_BUCKET giant-image

# When push is done...
docker rmi giant-image:latest

# Test pull
godep go run main.go pull s3://dogestry-test giant-image:latest
```

**Comment by dselans** [comment]
:+1:  :+1:  :+1: 

This should be dogestry-ng!

**Pull request merged**